unit2-apis-and-advanced-javascript/README.md

```diff
@@ -1,30 +1,114 @@
- - **Learning outcome**: Be able to discuss OOP interview questions (how to design games like chess, model animal species, etc.) and understand how to discuss the talking points for non-JS OOP 
-3. APIs via AJAX and JSON
- - jQuery's AJAX methods (ajax, get, post) & under-the-hood JavaScript's pure AJAX methods (only for educational reasons)
- - Success and failure handlers in callbacks
- - Firebase (or some other competitor) configuration and consumption
- - **Learning outcome**: Update the tic-tac-toe implementation from Unit 1 to store game state in Firebase, maintain leaderboards, have user accounts that users can log in to (insecurely) from any client
- - **Learning outcome**: Consume an existing API (http://jsonplaceholder.typicode.com/ or http://pokeapi.co/) and allow for the client to control different queries on the API which display different results
-4. Client Side UX Frameworks
- - React.js
- - Redux
- - d3.js
- - **Learning outcome**: Given a d3.js component (we could build it for them), design and implement a react.js app that has many d3 graphs whose state is managed by Redux
+- Building frontend javascript with Gulp: 2 days
+ - What is Gulp?
+ - How does Gulp transform the way we write frontend javascript? What does the file structure look like?
+ - Set up your first gulpfile
+ - How are tasks added to gulp? How are tasks composed with gulp?
+ - **Learning outcome**: Create a new project folder set up with Gulp and a gulpfile.
+ - **Learning outcome**: Add tasks to gulpfile such as: browserify, uglify, live-reload, babel (optional)
+- Multiline, Multifile JavaScript apps: 2 days
+ - How to leverage multiple packages (either from NPM or self created) to write maintainable code
+ - **Learning outcome**: Take the tic-tac-toe implementation from the previous unit and refactor into smaller modules in multiple files. The new modules should be self contained, reusable and extensible.
+ - **Learning outcome**: Working with a partner (or partners), come up with a Single webpage project idea (or use one from last unit: (Madlibs, Hangman, Math Quiz, Choose your own adventure game)). Plan out project into npm-style modules, owners, and deliverables (ie: for tic-tac-toe, Person A is responsible to writing the modules and listen for user input, Person B builds out the turn manager in the game, and so on). Work together using GitHub to develop and deploy your project. Obviously, you use gulp to build your js files for use on the frontend.
+
+### Week 2
+
+- Scope: .5 day
+ - Hoisting 
+ - Investigating the scope chain
+ - **Learning outcome**: Be able to answer common "what does this code output" tricky JS questions involving hoisting, scope. Examples can be seen here: https://www.sitepoint.com/5-typical-javascript-interview-exercises/ 
+- Closures: .5 day
+ - What are closures?
+ - How can they be used and when are they useful?
+ - **Learning outcome**: Solve tricky problems involving closures, examples:
+
+```js
```

Comment:
I think this is too much detail to this file. Instead put this in a different (assessment, assignment, whatever) and link to it from here.